File: src/LC366/Q3.java
package LC366;

import java.util.*;

public class Q3 {

    public static void main(String[] args) {
        String s1 = "101101";
        String s2 = "000000";
        System.out.println(util3(s1, s2, 6));
    }

    // dp
    public static int util3(String s1, String s2, int x) {
        hm = new HashMap<>();
        return helper(s1.toCharArray(), s2, 0, 0, false, x);
    }

    static HashMap<String, Integer> hm;

    public static int helper(char arr[], String arr2, int index, int cost, boolean op1, int x) {
        if (index == arr.length) {
            if (op1 == false && arr.toString().equals(arr2)) {
                return cost;
            } else {
                return Integer.MAX_VALUE;
            }
        }
        String key = arr.toString() + " " + op1;
        if (hm.containsKey(key)) {
            return hm.get(key);
        }
        int ans = Integer.MAX_VALUE;
        if (op1) {
            if (arr[index] == '0') {
                arr[index] = '1';
            } else {
                arr[index] = '0';
            }
            int ans1 = helper(arr, arr2, index + 1, cost + x, false, x);
            if (arr[index] == '0') {
                arr[index] = '1';
            } else {
                arr[index] = '0';
            }
            ans = ans1;
        } else {
            if (arr[index] == '0') {
                arr[index] = '1';
            } else {
                arr[index] = '0';
            }
            int ans1 = helper(arr, arr2, index + 1, cost, true, x);
            if (arr[index] == '0') {
                arr[index] = '1';
            } else {
                arr[index] = '0';
            }
            ans = ans1;
        }
        if (index < arr.length - 1) {
            if (arr[index] == '0') {
                arr[index] = '1';
            } else {
                arr[index] = '0';
            }
            if (arr[index + 1] == '0') {
                arr[index + 1] = '1';
            } else {
                arr[index + 1] = '0';
            }
            int ans2 = helper(arr, arr2, index + 2, cost + 1, op1, x);
            if (arr[index] == '0') {
                arr[index] = '1';
            } else {
                arr[index] = '0';
            }
            if (arr[index + 1] == '0') {
                arr[index + 1] = '1';
            } else {
                arr[index + 1] = '0';
            }
            ans = Math.min(ans, ans2);
        }

        int ans3 = helper(arr, arr2, index + 1, cost, op1, x);
        ans = Math.min(ans, ans3);
        hm.put(key, ans);
        return ans;
    }

    public static int util2(String s1, String s2, int x) {
        char arr[] = s1.toCharArray();
        char arr2[] = s2.toCharArray();
        boolean add = false;
        int prev = -1;
        int ans = 0;
        for (int i = 0; i < arr.length; i++) {
            if (arr[i] != arr2[i]) {
                if (add) {
                    if (prev == i - 1) {
                        ans += 1;
                    } else {
                        ans += x;
                    }
                    add = false;
                } else {
                    add = true;
                    prev = i;
                }
            }
        }
        if (!add) {
            return ans;
        } else {
            return -1;
        }
    }

    public static int util1(String s1, String s2, int x) {
        char arr[] = s1.toCharArray();
        char arr2[] = s2.toCharArray();
        boolean add = false;
        int ans = 0;
        for (int i = 0; i < arr.length; i++) {
            if (arr[i] != arr2[i]) {
                if (add) {
                    ans += x;
                    add = false;
                } else {
                    add = true;
                }
            }
        }
        if (!add) {
            return ans;
        } else {
            return -1;
        }
    }
}

File: LC118/Q3.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var minimumCoins = function (prices) {
  let dp = new Array(prices.length).fill(0);
  //base case
  dp[dp.length - 1] = prices[prices.length - 1];
  for (let i = prices.length - 2; i >= 0; i--) {
    let ans1 = dp[i + 1];
    let ans2 = dp[i + i] + 1 || 0;
    dp[i] = Math.min(ans1, ans2) + prices[i];
  }
  return dp[0];
};

//try 1-d dp

let map;

/**
 *
 * @param {number[]} prices
 * @param {number} index
 * @param {number} isFree
 * @param {number} costTillNow
 */
let util = (prices, index, freeNo, costTillNow) => {
  if (index == prices.length) {
    return costTillNow;
  }
  let key = index;
  if (Object.hasOwn(map, key)) {
    return map.key;
  }
  let ans = util(prices, index + 1, index + 1, costTillNow + prices[index]);
  if (freeNo > 0) {
    let ans2 = util(prices, index + 1, freeNo - 1, costTillNow);
    ans = Math.min(ans, ans2);
  }
  map.key = ans;
  return ans;
};
// let prices = [
//   176, 391, 307, 836, 76, 650, 625, 648, 466, 990, 643, 399, 422, 604, 328, 343,
//   835, 153, 92, 171, 802, 731, 483, 469, 27, 307, 194, 427, 824, 503, 661, 5,
//   540, 761, 127, 579, 324, 997, 891, 809, 968, 855, 397, 730, 583, 549, 493,
//   471, 809, 860, 771, 502, 177, 25, 107, 26, 321, 796, 726, 519, 800, 764, 804,
//   50, 471, 812, 747, 862, 220,
// ];
//console.log(minimumCoins(prices));

let prices = [1, 10, 1, 1];
console.log(minimumCoins(prices));
let arr = [1, 2];
//console.log(arr[3]||0);

// let obj = {
//     aloo:'a',
// };
// console.log(obj.aloo);
